### PR TITLE
Attaching to existing open zli connections [CWC-667]

### PR DESCRIPTION
**Pull request by tflpd**
## Description of the change

This adds functionality to manage zli open connections. This includes three new zli commands `list-connections` which lists any open zli connections, `attach` which lets the user re-attach to an existing open zli connection and `close` which lets the user close an existing open zli connection.

Please make sure you test this against both key-splitting enabled and key-splitting disabled agents, since the flow is different there might exist bugs on the one that do not exist on the other.

To test this:
- Open a new connection to a target with zli
- Type `exit` to exit that connection. Note that this exit does not close anymore the connection
- Using `lc` (list-connections`) get a list of any open zli connections
- Using `attach` reconnect to any of those. They should work as you never exited
- "Stress test" it by attaching to the same connection by different zli/webapp clients

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Refactor (code change with no functionality change)
- [ ] Enhancement (minor change below the level of a feature)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Remove feature (removes a feature we don't support anymore)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related JIRA tickets

Relates to JIRA: CWC-667

## Potential Security Impacts

For non-keysplitting agents, attaching to an existing connection will not remove any other attached clients. This is not a real threat but it should still be noted.

### Development

- [ ] Does the code changed/added as part of this pull request have test coverage?
- [ ] Do all tests related to the changed code pass in development?
- [ ] Have you ensured that at least one other person has tested your code?
- [X] Have you tested the code?
- [ ] Have you attached any pictures (if relevant)?

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -128,6 +128,26 @@ export function getTableOfTargets(targets: TargetSummary[], envs: EnvironmentDet
     return table.toString();
 }
 
+export function getTableOfConnections(connections: ConnectionSummary[]) : string
+{
+    const connIdLength = max(connections.map(c => c.id.length).concat(16));
```
Comment (Nabeelperson) [inline src/utils.ts:133]: This will always be 36 since they are all GUID are 36 characters.

File: src/utils.ts [hunk 2]
```diff
@@ -128,6 +128,26 @@ export function getTableOfTargets(targets: TargetSummary[], envs: EnvironmentDet
...
+    const header: string[] = ['Connection ID'];
+    const columnWidths = [connIdLength + 2];
+
+    // ref: https://github.com/cli-table/cli-table3
```
Comment (Nabeelperson) [inline src/utils.ts:137]: Delete duplicate ref

File: src/utils.ts [hunk 3]
```diff
@@ -128,6 +128,26 @@ export function getTableOfTargets(targets: TargetSummary[], envs: EnvironmentDet
...
+    connections.forEach(connection => {
+        const row = [connection.id];
+
+        table.push(row);
+    }
+    );
+
+    return table.toString();
+
+}
```
Comment (Nabeelperson) [inline src/utils.ts:149]: Please adjust this whitespace.

File: src/terminal/terminal.ts [hunk 4]
```diff
@@ -26,26 +25,24 @@ export class ShellTerminal implements IDisposable
     private terminalRunningStream: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(true);
     public terminalRunning: Observable<boolean> = this.terminalRunningStream.asObservable();
 
-    constructor(private logger: Logger, private configService: ConfigService, private connectionId: string, private parsedTarget: ParsedTargetString)
+    constructor(private logger: Logger, private configService: ConfigService, private connectionId: string, private targetType: TargetType, private serverId: string)
```
Comment (Nabeelperson) [inline src/terminal/terminal.ts:28]: Why did you remove `parsedTarget` here? It contains all the fields you are now explicitly passing in.

File: src/http.service/http.service.types.ts [hunk 5]
```diff
@@ -77,6 +77,7 @@ export interface ListSsmTargetsRequest
 }
 
 export interface SsmTargetSummary {
+    // The following is serverId
```
Comment (Nabeelperson) [inline src/http.service/http.service.types.ts:80]: I don't think this is a necessary comment.

File: src/handlers/create-shell.handler.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
+import { ConfigService } from '../../src/config.service/config.service';
+import { Logger } from '../../src/logger.service/logger';
+import { ShellTerminal } from '../../src/terminal/terminal';
+import { TargetType } from '../types';
+import { cleanExit } from './clean-exit.handler';
+import termsize from 'term-size';
+
+export async function createShellHandler(
```
Comment (Nabeelperson) [inline src/handlers/create-shell.handler.ts:8]: `createAndRunShell()` is probably a more appropriate name for this, as this is not a cli-driver handler method.

File: src/handlers/close-connection.handler.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import { ConfigService } from '../../src/config.service/config.service';
+import { ConnectionService } from '../../src/http.service/http.service';
+import { ConnectionState } from '../../src/http.service/http.service.types';
+import { Logger } from '../../src/logger.service/logger';
+import { cleanExit } from './clean-exit.handler';
+
+export async function closeConnectionHandler(
```
Comment (Nabeelperson) [inline src/handlers/close-connection.handler.ts:7]: Make sure to check if the connectionId is from the cli-space, otherwise you should not be closing the connection.

File: src/cli-driver.ts [hunk 8]
```diff
@@ -322,6 +325,42 @@ export class CliDriver
...
+            .command(
+                ['list-connections', 'lc'],
+                'List all open connections',
+                () => {},
+                async () => {
+                    await listConnectionsHandler(this.configService, this.logger);
+                }
+            )
+            .command(
+                'attach <connectionId>',
```
[the remark this answers is not in the document]
Comment (Nabeelperson) [reply src/cli-driver.ts:337]: You should have a regex check before passing in the connectionId to the handler, to make sure the argument is well formed. 

Probably this regex check should be made into a function and used in both places
https://github.com/bastionzero/zli/blob/feat/shell-reattachment/src/utils.ts#L77

Similar should be done for close connection command.

File: src/utils.ts (hunk 3 above)
Comment (tflpd) [reply src/utils.ts:149]: @lipmas Lint was not detecting this as an issue, should we add a rule?

File: src/terminal/terminal.ts (hunk 4 above)
Comment (tflpd) [reply src/terminal/terminal.ts:28]: This was needed because when attaching we do not have a `parsedTarget` struct to pass here, we have only the `connectionId` and the `targetType`

File: src/handlers/list-connections.handler.ts [hunk 9]
```diff
@@ -1,15 +1,19 @@
...
 
 export async function listConnectionsHandler(
     configService: ConfigService,
-    logger: Logger
+    logger: Logger,
+    ssmTargets: Promise<TargetSummary[]>,
+    sshTargets: Promise<TargetSummary[]>
 ){
+    // await and concatenate
+    const allTargets = [...await ssmTargets, ...await sshTargets];
```
Comment (Nabeelperson) [inline src/handlers/list-connections.handler.ts:16]: You should probably only await on this after checking the length of openConnections.

File: src/utils.ts [hunk 10]
```diff
@@ -128,21 +133,19 @@ export function getTableOfTargets(targets: TargetSummary[], envs: EnvironmentDet
...
+    const columnWidths = [targetNameLength + 2, connIdLength + 2, 20];
 
-    // ref: https://github.com/cli-table/cli-table3
     const table = new Table({ head: header, colWidths: columnWidths });
-
+    const dateOptions = {year: '2-digit', month: 'numeric', day: 'numeric', hour:'numeric', minute:'numeric', hour12: true};
     connections.forEach(connection => {
-        const row = [connection.id];
-
+        const row = [allTargets.filter(t => t.id == connection.serverId).pop().name, connection.id, new Date(connection.timeCreated).toLocaleString('en-US', dateOptions as any)];
```
Comment (Nabeelperson) [inline src/utils.ts:146]: Does mike agree about the order of information here?

**Review by Nabeelperson — COMMENTED** [review]
Are we including a `--json` flag for list-connections?

File: src/utils.ts (hunk 10 above)
Comment (tflpd) [reply src/utils.ts:146]: Yes

File: src/handlers/attach.handler.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+import { createAndRunShell } from '../../src/shell-utils';
+
+export async function attachHandler(
+    configService: ConfigService,
+    logger: Logger,
+    connectionId: string
+){
+    const connectionService = new ConnectionService(configService, logger);
+    const connectionSummary = await connectionService.GetConnection(connectionId);
+    if (connectionSummary.state !== ConnectionState.Open){
```
Comment (lipmas) [inline src/handlers/attach.handler.ts:15]: Also add a check here for connection belonging to cli-space?

File: src/shell-utils.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+    // space names are not unique, make sure to find the latest active one
+    const cliSpace = listSessions.sessions.filter(s => s.displayName === 'cli-space' && s.state == SessionState.Active); // TODO: cli-space name can be changed in config
+
+    // maybe make a session
+    let cliSessionId: string;
+    if(cliSpace.length === 0) {
+        cliSessionId =  await sessionService.CreateSession('cli-space');
+    } else {
+        // there should only be 1 active 'cli-space' session
```
Comment (lipmas) [inline src/shell-utils.ts:80]: can we log a warning if there is more than 1 active cli-space

File: src/handlers/list-connections.handler.ts [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    const cliSpace = listSessions.sessions.filter(s => s.displayName === 'cli-space' && s.state == SessionState.Active); // TODO: cli-space name can be changed in config
+
+    let cliSessionId: string;
+    if(cliSpace.length === 0) {
+        logger.info('There is no cli session available');
+        await cleanExit(0, logger);
+    } else {
+        // there should only be 1 active 'cli-space' session
+        cliSessionId = cliSpace.pop().id;
+    }
```
Comment (lipmas) [inline src/handlers/list-connections.handler.ts:30]: use common `getCliSpaceId` function here

Comment (tflpd) [reply src/handlers/list-connections.handler.ts:30]: I should not. There is a small difference. The `getCliSpaceId` creates a session if there is none available. However in here we do not want to create one, we want to exit if none is available

Comment (lipmas) [reply src/handlers/list-connections.handler.ts:30]: Ah i see ok. Then maybe you could just split `getCliSpaceId` and `createCliSpace` into 2 separate functions so that we could reuse the `getCliSpaceId` code here.

File: src/handlers/middleware.handler.ts [hunk 14]
```diff
@@ -48,11 +51,14 @@ export function fetchDataMiddleware(configService: ConfigService, logger: Logger
 
     const envs = envService.ListEnvironments();
 
+    const cliSpaceId = getCliSpaceId(sessionService, logger);
```
Comment (lipmas) [inline src/handlers/middleware.handler.ts:54]: I dont think we should be continuing to use this patten of pushing data retrieval to this middleware. This increases the number of API calls for every single zli command. We are already starting to see some serious performance issues when running stress tests where many concurrent zli commands are run and each one currently results in 4 API calls just from this middleware layer.

Ideally we should load just the data that is needed in the beginning of each command handler. We can have the data retrieval functions be in common util methods and have each handler load what data it needs in parallel by combining all the methods in a single `await Promise.All()` instead of running all of them in series. 

cc @Nabeelperson If you agree we can open up another issue to refactor the existing code in a separate PR but lets at least not add more middleware API calls here.

File: src/terminal/terminal.ts [hunk 15]
```diff
@@ -97,16 +93,19 @@ export class ShellTerminal implements IDisposable
                 case ShellEventType.Start:
                     this.blockInput = false;
                     this.terminalRunningStream.next(true);
+                    // Trigger resize to force the terminal to refresh the output
+                    const tempTerminalSize : TerminalSize = {rows: this.currentTerminalSize.rows + 1, columns: this.currentTerminalSize.columns + 1};
+                    if(! this.blockInput)
```
Comment (lipmas) [inline src/terminal/terminal.ts:98]: nit: dont need this if because we set `this.blockInput = false;` on line 94

File: src/handlers/middleware.handler.ts (hunk 14 above)
Comment (Nabeelperson) [reply src/handlers/middleware.handler.ts:54]: Yes, I agree we should move away from the greedy fetching. Yes lets open up another issue. I don't mind leaving this code for cliSpaceId where it is for now though since the refactor is coming anyways.

**Comment by Nabeelperson** [comment]
For the `--json` flag, when there are no connections to return, it should return an empty json.

File: src/handlers/middleware.handler.ts (hunk 14 above)
Comment (lipmas) [reply src/handlers/middleware.handler.ts]: created #114
